File: JAVA/CircularQueue.java
// Description 
// A circular queue is a data structure that uses a fixed-size array to store elements,
// and it supports both enqueue (add elements) and dequeue (remove elements) operations.
// The circular queue overcomes the limitation of a regular queue, where the queue becomes
// full after a certain number of elements are added, making it inefficient for long-running applications.

// Implementation

public class CircularQueue {
    private int[] elements;
    private int front;  // Index of the front element
    private int rear;   // Index of the rear element
    private int size;   // Current number of elements
    private int capacity; // Maximum capacity of the queue

    public CircularQueue(int capacity) {
        this.capacity = capacity;
        this.elements = new int[capacity];
        this.front = 0;
        this.rear = -1;  // Initially, the queue is empty
        this.size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public int size() {
        return size;
    }

    public void enqueue(int item) {
        if (isFull()) {
            System.out.println("Queue is full. Cannot enqueue.");
            return;
        }

        rear = (rear + 1) % capacity; // Wrap around if the rear pointer goes beyond the array size
        elements[rear] = item;
        size++;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is empty. Cannot dequeue.");
            return -1; // You can choose a different error code or handle it differently
        }

        int item = elements[front];
        front = (front + 1) % capacity; // Wrap around if the front pointer goes beyond the array size
        size--;
        return item;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty.");
            return -1; // You can choose a different error code or handle it differently
        }
        return elements[front];
    }

    public void display() {
        if (isEmpty()) {
            System.out.println("Queue is empty.");
            return;
        }

        int current = front;
        for (int i = 0; i < size; i++) {
            System.out.print(elements[current] + " ");
            current = (current + 1) % capacity;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        CircularQueue queue = new CircularQueue(5);

        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);
        queue.enqueue(4);
        queue.enqueue(5);

        System.out.print("Queue elements: ");
        queue.display();

        System.out.println("Dequeued item: " + queue.dequeue());
        System.out.println("Dequeued item: " + queue.dequeue());

        System.out.print("Queue elements after dequeuing: ");
        queue.display();

        queue.enqueue(6);
        queue.enqueue(7);

        System.out.print("Queue elements after enqueuing: ");
        queue.display();
    }
}

File: C++/CircularQueue.cpp
#include <iostream>
using namespace std;

const int MAX_SIZE = 5; // Maximum size of the queue

class CircularQueue {
private:
    int front, rear;
    int arr[MAX_SIZE];

public:
    CircularQueue() {
        front = rear = -1; 
    }

    // check if the queue is empty
    bool isEmpty() {
        return (front == -1 && rear == -1);
    }

    //  check if the queue is full
    bool isFull() {
        return (front == (rear + 1) % MAX_SIZE);
    }

    //  enqueue  an element
    void enqueue(int value) {
        if (isFull()) {
            cout << "Queue is full. Cannot enqueue." << endl;
            return;
        }
        if (isEmpty()) {
            front = rear = 0; // 
        } else {
            rear = (rear + 1) % MAX_SIZE; // Circular increment of rear
        }
        arr[rear] = value;
        cout << value << " enqueued to the queue." << endl;
    }

    // dequeue  an element
    void dequeue() {
        if (isEmpty()) {
            cout << "Queue is empty. Cannot dequeue." << endl;
            return;
        }
        cout << arr[front] << " dequeued from the queue." << endl;
        if (front == rear) {
            front = rear = -1; 
        } else {
            front = (front + 1) % MAX_SIZE; // Circular increment of front
        }
    }

    // display the elements of the queue
    void display() {
        if (isEmpty()) {
            cout << "Queue is empty." << endl;
            return;
        }
        cout << "Elements in the queue: ";
        int i = front;
        while (true) {
            cout << arr[i] << " ";
            if (i == rear)
                break;
            i = (i + 1) % MAX_SIZE; // Circular increment of index
        }
        cout << endl;
    }
};

int main() {
    CircularQueue queue;
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    queue.display();
    queue.dequeue();
    queue.display();
    queue.enqueue(4);
    queue.enqueue(5);
    queue.display();
    queue.enqueue(6); // This will show an error as the queue is full.
    return 0;
}
